File: src/main/java/com/vaadin/example/ui/SearchPanel.java
package com.vaadin.example.ui;

import com.vaadin.example.utils.FileUtils;
import com.vaadin.example.search.MdProcessor;
import com.vaadin.example.search.Searcher;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.H1;
import com.vaadin.flow.component.html.Input;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;

import java.util.ArrayList;
import java.util.List;

/**
 * Панель с инструментами поиска и вывода результатов
 */
public class SearchPanel extends VerticalLayout {
    private Searcher searcher = new Searcher();
    private List<DescriptionItem> resultItems = new ArrayList<>();
    private VerticalLayout textLayout = new VerticalLayout();
    private VerticalLayout viewer = new VerticalLayout();
    private FilterCheckboxes filterCheckboxes = new FilterCheckboxes();

    public SearchPanel() {
        setSizeFull();
        setAlignItems(Alignment.CENTER);
        H1 title = new H1("Поиск документов");
        title.setClassName("lumo-success-text-color");
        add(title);
        HorizontalLayout buttonLayout = new HorizontalLayout();
        buttonLayout.addClassName("searchPanel");
        add(buttonLayout);
        Input filterString = new Input();
        Button button = new Button("Найти",
                event -> search(filterString.getValue()));
        button.addClassName("searchButton");
        buttonLayout.add(filterString);
        buttonLayout.setWidth("100%");
        filterString.setWidth("90%");
        filterString.setHeight("30px");
        buttonLayout.setAlignItems(Alignment.CENTER);
        buttonLayout.add(button);
        add(filterCheckboxes);
        HorizontalLayout resultLayout = new HorizontalLayout();
        add(resultLayout);
        resultLayout.add(textLayout);
        resultLayout.addAndExpand(viewer);
        resultLayout.setSizeFull();
        textLayout.addClassName("resultLayout");
        setHeight("100%");
    }

    /**
     * Запустить поиск по документам
     * @param searchText текст для поиска
     */
    private void search(String searchText) {
        viewer.removeAll();
        resultItems.forEach(l -> textLayout.remove(l));
        resultItems.clear();
        List<String> results = searcher.getMatchesFilenames(searchText);
        results.forEach(r -> {
            String fileData = FileUtils.getFileData("md/" + r);
            String category = MdProcessor.getCategory(fileData);
            String resourceFilename = MdProcessor.getResourceFilename(fileData);
            DescriptionItem item = new DescriptionItem(/*Categories.valueOf(category)*/null, MdProcessor.getTitle(fileData), //todo resolve encoding bug
                    MdProcessor.getContent(fileData), resourceFilename, viewer);
            resultItems.add(item);
            textLayout.add(item);
        });
    }
}

File: src/main/kotlin/com/vaadin/example/ui/SearchPanel.kt
package com.vaadin.example.ui

import com.vaadin.example.search.Categories
import com.vaadin.example.search.MdProcessor
import com.vaadin.example.search.Searcher
import com.vaadin.example.utils.FileUtils
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.html.H1
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField
import kotlin.collections.HashMap

class SearchPanel : VerticalLayout() {
    private val searcher = Searcher()
    private val resultItems : MutableList<DescriptionItem> = mutableListOf()
    private val textLayout = VerticalLayout()
    val viewer = VerticalLayout()
    private val filterCheckboxes = FilterCheckboxes(this)
    private var results = emptyList<String>()

    init {
        setSizeFull()
        alignItems = FlexComponent.Alignment.CENTER
        val title = H1("Поиск документов")
        add(title)
        val buttonLayout = HorizontalLayout()
        add(buttonLayout)
        val filterString = TextField()
        filterString.addKeyPressListener { event ->
            if (event.key.matches("Enter")) search(filterString.value) }
        val button = Button("Найти") {search(filterString.value)}
        buttonLayout.add(filterString)
        buttonLayout.setWidthFull()
        buttonLayout.alignItems = FlexComponent.Alignment.CENTER
        buttonLayout.add(button)
        add(filterCheckboxes)
        val resultLayout = HorizontalLayout()
        add(resultLayout)
        resultLayout.add(textLayout)
        resultLayout.addAndExpand(viewer)
        resultLayout.setSizeFull()
        textLayout.setFlexGrow(1.0, filterString)
        textLayout.setWidthFull()
        setHeightFull()
    }

    fun search(searchText : String) {
        results = searcher.getMatchesFilenames(searchText)

        viewer.removeAll()
        resultItems.clear()

        val foundCategories = HashMap<Categories, Int>()
        results.forEach{
            val fileData = FileUtils.getFileData("md/$it")
            val categoryName = MdProcessor.getCategory(fileData)
            val category = Categories.getByLabel(categoryName)
            val resourceFilename = MdProcessor.getResourceFilename(fileData)
            val item = DescriptionItem(category, MdProcessor.getTitle(fileData), MdProcessor.getContent(fileData),
                    resourceFilename, viewer)
            resultItems.add(item)
            if (category != null) {
                val itemCount : Int? = foundCategories[category]
                if (itemCount != null)
                    foundCategories[category] = itemCount
                else
                    foundCategories[category] = 1
            }
        }
        filterCheckboxes.refresh(foundCategories)
        showItems(emptySet())
    }

    fun showItems(filter : Set<Categories>) {
        viewer.removeAll()
        resultItems.forEach { l -> textLayout.remove(l) }
        resultItems.forEach {
            if (filter.isEmpty() || filter.contains(it.category))
                textLayout.add(it)
        }
    }
}